### ISSUE TITLE
feature request: track "exact" types (excluding subclasses)

### ISSUE THREAD

**Issue by goodspark** [astral-sh/ty#2432]
### Summary

Playground: https://play.ty.dev/fddded41-5c96-4852-9849-bd44cdf03074

```py
from datetime import datetime, timedelta
from dataclasses import dataclass

@dataclass
class Obj:
    x: datetime

a: Obj | None = Obj(datetime.now())
# Since a is always truthy, the boolean statement continues and returns datetime
b = (a and a.x - timedelta(days=1))
# c should always be datetime here, so the diagnostic is incorrect
c = datetime.now() + timedelta(days=1) > b
# Operator `>` is not supported between objects of type `datetime` and `(Obj & ~AlwaysTruthy) | datetime`
```

If `a` was just `datetime | None`, this would work fine, but this use case is meaningful as often container data types might be None (ex. a DB row lookup).

To be fair, this diagnostic is useful _if `a` was not AlwaysTruthy, such as through a function call (which is more realistic)_. However, the diagnostic message is misleading as it currently is. One could argue instead we should write code using an if or ternary expression.

```py
b = None
if a:
    b = a.x - timedelta(days=1)
# or ...
b = None if a is None else a.x - timedelta(days=1)
```

With this change, ty will correctly infer `b` is always datetime and have no issue. And it will also correctly error with an accurate message if `a` was not AlwaysTruthy (ie. None can't be compared to datetime).

---

Terms used to find already-reported issues:
- `unsupported-operation boolean`
- `unsupported-operation and`
- `unsupported-operation truthy`

#1972 might be a match. It talks about types being constrained by values. In a way, the type of `b` does get constrained by the resulting value of evaluating the compound boolean expression. But my example is not using a direct value but instead the resulting value of an operation.

### Version

0.0.11

**Comment by carljm** [comment]
Thanks for the report!

ty does not consider `Obj` to be an always-truthy type, because there could be subclasses of it which define `__bool__` to sometimes return `False`. Instances of such subclasses would not necessarily be truthy, and are part of the type `Obj`.

Of course in this case we could in theory tell that `a` cannot be an instance of such a subclass, because we directly constructed an `Obj` instance just above. But making this distinction would require ty to track "exact" types (instance types that _don't_ include subclass instances). This would add a lot of complexity, and the usefulness of it is questionable, since it can only apply in limited scenarios, locally right after the construction of an instance from a concrete class.

Another workaround for this case would be to mark `Obj` as `@final` -- then ty will treat it as always truthy.

We can keep this open as a feature request for exact types, but I don't think that will be a priority in the near term.

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-fable-5 on behalf of David.

Another real-world instance of this, from running ty 0.0.72 over [pydantic-ai](https://github.com/pydantic/pydantic-ai) — 4 diagnostics, all the same `x and x.attr` shape, and here the leftover `& ~AlwaysTruthy` arm surfaces as an `invalid-argument-type` rather than an operator error:

```py
class Function:
    name: str | None = None


def handle(name: str | None) -> None: ...


def f(function: Function | None) -> None:
    handle(function and function.name)
```

```console
$ uvx ty check repro.py
repro.py:9:12: error[invalid-argument-type] Argument to function `handle` is incorrect: Expected `str | None`, found `(Function & ~AlwaysTruthy) | None | str`
Found 1 diagnostic
```

pyright is clean (it eliminates the falsy arm for a plain class with no `__bool__`/`__len__`); mypy also keeps the class in the union, so ty isn't alone here — but pyright's answer is the one that matches runtime.

The pydantic-ai sites are provider streaming adapters, where the SDK hands us an optional sub-object and we forward two of its fields, e.g. [`models/groq.py`](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/models/groq.py): `tool_name=dtc.function and dtc.function.name`. `x and x.attr` seems to be the common spelling that runs into this — worth noting alongside the `a and a.x - timedelta(...)` example in the issue body, since it means the pattern shows up wherever an optional object's attribute is forwarded, not just in arithmetic.

Notes: ty 0.0.72 via `uvx ty check`.

**Comment by carljm** [comment]
@dsfaccini I consider this a true-positive from ty (and mypy), and a false-negative from pyright. You say pyright's result "matches the runtime", but that's only true if you ignore the possibility that a subclass of `Function` could override `__bool__` to sometimes return `False`. If you want to rule out that possibility, decorate `Function` with `@final`. If you want to make that possibility irrelevant, write more precise code and use `function.name if function is not None else function`, which precisely describes the desired "is not None" condition instead of conflating it with truthiness of `Function`.

Alternatively a case could be made that we should just (unsoundly) ignore the possibility of subclass `__bool__` overrides, because it is usually practically unlikely. I don't think we currently have an issue open for that -- if you'd like to argue for that behavior, feel free to open one! (It's separate from this issue, which is about tracking "exact" types, something that is not involved in your example.)

**Comment by dsfaccini** [comment]
@carljm I appreciate your reply and your guidance! you were indeed right, I'm making changes on our side to get this off my list

sorry if the comment on this issue was a bit off topic, I'm desperate so I'm shotgunning all the gaps with claude/codex to see if we can adopt ty soon

**Comment by carljm** [comment]
@dsfaccini No worries! Really appreciate your interest in ty and your feedback from a real-world codebase, keep it coming!
